common/src/main/java/org/conscrypt/ConscryptEngineSocket.java

```diff
@@ -446,18 +448,28 @@ public final void close() throws IOException {
             stateLock.notifyAll();
         }
 
+        // Close the engine.
         try {
-            // Close the underlying socket.
-            super.close();
-        } finally {
-            // Close the engine.
             engine.closeInbound();
             engine.closeOutbound();
-            
-            // Release any resources we're holding
-            if (in != null) {
-                in.release();
+            // Closing the outbound direction of a connected engine will trigger a TLS close
+            // notify, which we should try and send.
+            // If we don't, then closeOutbound won't be able to free resources because there are
+            // bytes queued for transmission so drain the queue those and call closeOutbound a
+            // second time.
+            if (previousState >= STATE_HANDSHAKE_STARTED && out != null) {
+                drainOutgoingQueue();
+                engine.closeOutbound();
             }
+        } finally {
+            // In case of an exception thrown while closing the engine, we still need to close the
+            // underlying socket.
+            super.close();
+        }
+
+        // Release any resources we're holding
+        if (in != null) {
+            in.release();
```

Comment:
Doesn't the allocated buffer need to be freed in the exceptional case described on 465?

Reply:
Yes indeed it does - good catch, thanks!

Reply:
I think we may want to nest another try/finally around super.close because it declares `throws IOException`.

Reply:
Doh! Thanks again, perils of late night coding.